``pylint`` equivalent to ``pylint .``

Pierre-Sassoulas opened the pull request:
## Type of Changes

<!-- Leave the corresponding lines for the applicable type of change: -->

|     | Type                   |
| --- | ---------------------- |
| ✓   | :sparkles: New feature |

## Description

``pylint`` is now equivalent to ``pylint .`` and won't show the help by default anymore.
The help is still available with ``pylint --help`` or ``pylint --long-help``.

Closes #5701

Pierre-Sassoulas committed:
[cleanup] Remove behavior that is in argparse already

```
    def parse_known_args(self, args=None, namespace=None):
        if args is None:
            # args default to the system args
            args = _sys.argv[1:]
        else:
            # make sure that args are mutable
            args = list(args)
```

Pierre-Sassoulas committed:
Make 'pylint' equivalent to 'pylint .'

Closes #5701

Pierre-Sassoulas commented:
Hmm, we might want to have it in a beta for a somewhat longish time, this feel incompatible with releasing in a week. On one hand the change is not supposed to be complex, on the other hand it's not, because the option parsing and configuration parsing is very complex (there's no explicit positional argument like you would expect with argparse, we're just implicitly relying on the first args being the fileçor_module if we don't have the stdin option -- maybe other condition apply).

jacobtylerwalls commented on pylint/lint/pylinter.py:
```diff
@@ -658,13 +658,14 @@ def check(self, files_or_modules: Sequence[str]) -> None:
         if self.config.from_stdin:
             if len(files_or_modules) != 1:
-                raise exceptions.InvalidArgsError(
-                    "Missing filename required for --from-stdin"
-                )
+                print("Missing filename required for --from-stdin")
+                sys.exit(32)
```
Is this necessary for the current change? A small thing, but I think print() goes to stdout and exceptions go to stderr, so it might be nicer the way it was.

jacobtylerwalls replied:
Got it, makes sense.

jacobtylerwalls commented on pylint/lint/run.py:
```diff
@@ -176,10 +176,8 @@ def __init__(
             return
 
         # Display help if there are no files to lint or no checks enabled
```
May as well clean up this comment, as we're not displaying help (right?)

Pierre-Sassoulas replied:
44c098e73f044db59428aacdbad0e7f93501f748

jacobtylerwalls commented on tests/test_self.py:
```diff
@@ -491,8 +491,9 @@ def test_pylintrc_comments_in_values(self) -> None:
     def test_no_crash_with_formatting_regex_defaults(self) -> None:
+        path = join(HERE, "regrtest_data", "empty.py")
         self._runtest(
-            ["--ignore-patterns=a"], reporter=TextReporter(StringIO()), code=32
```
What was making the prior code emit 32?

Pierre-Sassoulas replied:
"No file to lint"

jacobtylerwalls commented on pylint/config/arguments_manager.py:
```diff
@@ -225,12 +225,9 @@ def _parse_command_line_configuration(
         self, arguments: Sequence[str] | None = None
     ) -> list[str]:
         """Parse the arguments found on the command line into the namespace."""
-        arguments = sys.argv[1:] if arguments is None else arguments
```
Should we remove the default `None`?

jacobtylerwalls commented on doc/whatsnew/fragments/5701.breaking:
```diff
@@ -0,0 +1,4 @@
+``pylint`` is now equivalent to ``pylint .`` and won't show the help by default anymore.
+The help is still available with ``pylint --help`` or ``pylint --long-help``.
```
Let's not promise equivalence with `pylint .`. I'm showing a behavior difference. Before, cd to an directory you haven't installed with pip, invoke `pylint .`, you get a quick lint of the `__init__.py` only; now you get the whole package linted.

Pierre-Sassoulas replied:
44c098e73f044db59428aacdbad0e7f93501f748

jacobtylerwalls reviewed (commented):
Thanks!

Wondering if you checked pyreverse?

Pierre-Sassoulas commented:
I'm thinking more and more that we should base it on https://github.com/pylint-dev/pylint/pull/7496/files so we can just add a default value for ``files`` instead of hacking the implicit list of args.

DanielNoord commented:
> I'm thinking more and more that we should base it on https://github.com/pylint-dev/pylint/pull/7496/files so we can just add a default value for `files` instead of hacking the implicit list of args.

I would be very much in favour of that, but need some direction in moving that PR forward.

Pierre-Sassoulas commented:
Let's move that change to 4.0.0 and try to add the ``files`` option in ``3.1.0``, I don't want to rush this.

Pierre-Sassoulas committed:
Address review feedback

Reword the 5701.breaking fragment to describe the actual behavior change
(no positional argument now lints the current directory) instead of
claiming equivalence with ``pylint .``. Fix the misleading "Display help"
comment in run.py — the code prints "No messages to check" and exits, it
does not display the help.
